### ISSUE TITLE
deep_supervision

### ISSUE THREAD

**Issue by long123524** [jeya-maria-jose/UNeXt-pytorch#7]
![image](https://user-images.githubusercontent.com/73996375/162610393-6f8de8e3-ad4b-42a0-b0d8-d8c590d29eaf.png)
I want to use deep_supervision, but I find a error, should I how to deal with this error? What the function of the deep_supervision? Thank you!

**Comment by Zhu1Li** [comment]
I finished the training on BUSI without DS, but when I tried to train on BUSI dataset with DS and met the same question, have you solved it? @long123524

**Comment by Zhu1Li** [comment]
> I finished the training on BUSI without DS, but when I tried to train on BUSI dataset with DS and met the same question, have you solved it? @long123524

Hey I got where the problem is, it happens because your "output"(torch.Size([1, 256, 256])) and "target"( torch.Size([8, 1, 256, 256])) is not dimmensional equal so your criterion function cannot calculate the right loss, all you need to do is changing the 
           for output in outputs:
                loss += criterion(output, target)
into
           for i in range(len(outputs)):
                loss += criterion(outputs[i], target[i])
in both train function(119,120 line) and validate function(166, 167 line) of train.py; then the code with configure "--deep_supervision True" will run correctly.

**Comment by Zhu1Li** [comment]
@jeya-maria-jose Thanks for your excellent work and could you please update this part of code so that it could be convinent for later people to debug UNeXt with deep supervision? 
Also, I want to ask a question about deep supervision, I found that either UNet++ or UNeXt, they reach convergence more slowly with deep supervision than without it,  for example I use BUSI and configures are as follows:
>--dataset BUSI --arch UNext --name UNext_BUSI_DS --img_ext .png --mask_ext .png --lr 0.0001 --epochs 400 --input_w 256 --input_h 256 --b 8
and without supervision, the val_iou can reach 0.6 easily within 30 epochs, but with deep supervision, the val_iou is still about 0.18 when all the 400 epochs end and I don't know what the problem is. Should I use larger learning rate or more epochs when training UNeXt with deep supervision than without? Or if there're other tricks to improve the performence please tell me. I would be appreciate for your help. Looking forward to your reply!

**Comment by Zhu1Li** [comment]
> > I finished the training on BUSI without DS, but when I tried to train on BUSI dataset with DS and met the same question, have you solved it? @long123524
> 
> Hey I got where the problem is, it happens because your "output"(torch.Size([1, 256, 256])) and "target"( torch.Size([8, 1, 256, 256])) is not dimmensional equal so your criterion function cannot calculate the right loss, all you need to do is changing the
> 
> > for output in outputs:
> > > loss += criterion(output, target)
> 
> into
> 
> > for i in range(len(outputs)):
> > > loss += criterion(outputs[i], target[i])
> 
> in both train function(119,120 line) and validate function(166, 167 line) of train.py; then the code with configure "--deep_supervision True" will run correctly.

Found something wrong with code, I used same configures but got a very terrible IoU score on BUSI dataset with deep supervision than without, and I compared the code of UNeXt and UNet++, found that **archs.py of UNeXt which defines the network architecture considers nothing about deep supervision(just one stage to output)**, while the losses.py and the part of  train&vaildate function about deep supervision in train.py is totall the same as UNet++'s code. 
So I guess that the author did not consider using deep supervision in the network, but just modified the code framework against UNet++ without removing the deep supervision part.

**Comment by jeya-maria-jose** [comment]
Yes, deep supervision part was not used in UNeXt training.

**Comment by Alan-Py** [comment]
> > I finished the training on BUSI without DS, but when I tried to train on BUSI dataset with DS and met the same question, have you solved it? @long123524
> 
> Hey I got where the problem is, it happens because your "output"(torch.Size([1, 256, 256])) and "target"( torch.Size([8, 1, 256, 256])) is not dimmensional equal so your criterion function cannot calculate the right loss, all you need to do is changing the
> 
> > for output in outputs:
> > > loss += criterion(output, target)
> 
> into
> 
> > for i in range(len(outputs)):
> > > loss += criterion(outputs[i], target[i])
> 
> in both train function(119,120 line) and validate function(166, 167 line) of train.py; then the code with configure "--deep_supervision True" will run correctly.

Good job
